File: analysis/scripts/compute_social_learning_outcomes.py
#!/usr/bin/env python3
"""Compute social learning group averages and errors."""

from __future__ import annotations

import argparse

from _common import group_by, number, read_records, write_records


def compute(records: list[dict], true_value: float | None) -> list[dict]:
    output = []
    for key, grouped in group_by(records, ["game_id", "group_id", "round"]).items():
        estimates = [number(row.get("action_value", row.get("estimate"))) for row in grouped]
        group_average = sum(estimates) / len(estimates) if estimates else 0
        error = abs(group_average - true_value) if true_value is not None else ""
        for row, estimate in zip(grouped, estimates):
            output.append(
                {
                    **row,
                    "estimate": estimate,
                    "group_average": group_average,
                    "distance_from_group_average": abs(estimate - group_average),
                    "absolute_error": error,
                    "game_id": key[0],
                    "group_id": key[1],
                    "round": key[2],
                }
            )
    return output


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("input")
    parser.add_argument("output")
    parser.add_argument("--true-value", type=float, default=None)
    args = parser.parse_args()
    write_records(args.output, compute(read_records(args.input), args.true_value))


if __name__ == "__main__":
    main()

File: analysis/scripts/compute_social_learning_outcomes.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("Usage: Rscript compute_social_learning_outcomes.R input.csv output.csv [true_value]")
}

data <- read.csv(args[[1]], stringsAsFactors = FALSE)
true_value <- if (length(args) >= 3) as.numeric(args[[3]]) else NA
keys <- c("game_id", "group_id", "round")
split_rows <- split(data, interaction(data[keys], drop = TRUE))

rows <- lapply(split_rows, function(group) {
  estimate <- as.numeric(if ("action_value" %in% names(group)) group$action_value else group$estimate)
  group_average <- mean(estimate, na.rm = TRUE)
  group$estimate <- estimate
  group$group_average <- group_average
  group$distance_from_group_average <- abs(estimate - group_average)
  group$absolute_error <- if (!is.na(true_value)) abs(group_average - true_value) else NA
  group
})

output <- do.call(rbind, rows)
dir.create(dirname(args[[2]]), recursive = TRUE, showWarnings = FALSE)
write.csv(output, args[[2]], row.names = FALSE)
